File: ios/Classes/FlutterUnifiedBannerView.m
//
//  FlutterUnifiedBannerView.m
//  adnet_qq
//
//  Created by Dolphin on 2019/7/7.
//

#import <Foundation/Foundation.h>
#import <Flutter/Flutter.h>
#import "FlutterUnifiedBannerView.h"
#import "GDTUnifiedBannerView.h"
#import "PluginSettings.h"
#import "AdnetQqPlugin.h"

@implementation FlutterUnifiedBannerView {
    int64_t _viewId;
    GDTUnifiedBannerView *bv;
    FlutterMethodChannel* _channel;
    NSString *posId;
}

- (instancetype)initWithFrame:(CGRect)frame viewId:(int64_t)viewId arguments:(id)args binaryMessenger:(NSObject<FlutterBinaryMessenger> *)messenger
{
    if([super init]) {
        NSDictionary *dict = args;
        posId = dict[@"posId"];
        _viewId = viewId;
        NSString *channelName = [NSString stringWithFormat:@"%@_%lld",[PluginSettings getInstance]->UNIFIED_BANNER_VIEW_ID,viewId];
        _channel = [FlutterMethodChannel methodChannelWithName:channelName binaryMessenger:messenger];
        __weak __typeof__(self) weakSelf = self;
        [_channel setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result){
            [weakSelf onMethodCall:call result: result];
        }];
        //bv = [[GDTUnifiedBannerView alloc] initWithFrame:frame appId:[PluginSettings getInstance]->APP_ID placementId:posId viewController:AdnetQqPlugin.getViewController];
    }
    return self;
}

- (nonnull GDTUnifiedBannerView *)getView{
    if(!bv) {
        CGRect rect = {CGPointZero, CGSizeMake(375, 60)};
        //NSLog(@"APP_ID:%@,posId:%@,viewController:%@",[PluginSettings getInstance] -> APP_ID,posId, [AdnetQqPlugin getViewController]);
        bv = [[GDTUnifiedBannerView alloc]
              initWithFrame:rect appId:[PluginSettings getInstance] -> APP_ID placementId: posId viewController:[AdnetQqPlugin getViewController]
              ];
        bv.animated = NO;
        bv.delegate = self;
    }
    return bv;
}

- (nonnull UIView *)view {
    return [self getView];
}

- (void)onMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result
{
    if([@"refresh" isEqual:call.method]) {
        [[self getView] loadAdAndShow];
        result(@(YES));
        return;
    }
    if([@"close" isEqual:call.method] && bv) {
        [bv removeFromSuperview];
        bv.delegate = nil;
        result(@(YES));
        return;
    }
}

/**
 *  请求广告条数据成功后调用
 *  当接收服务器返回的广告数据成功后调用该函数
 */
- (void)unifiedBannerViewDidLoad:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdReceived");
    [_channel invokeMethod:@"onAdRecieved" arguments:nil];
}

/**
 *  请求广告条数据失败后调用
 *  当接收服务器返回的广告数据失败后调用该函数
 */
- (void)unifiedBannerViewFailedToLoad:(GDTUnifiedBannerView *)unifiedBannerView error:(NSError *)error
{
    NSLog(@"onNoAd %@", error);
    [_channel invokeMethod:@"onNoAd" arguments:nil];
}

/**
 *  banner2.0曝光回调
 */
- (void)unifiedBannerViewWillExpose:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdExposure");
    [_channel invokeMethod:@"onAdExposure" arguments:nil];
}

/**
 *  banner2.0点击回调
 */
- (void)unifiedBannerViewClicked:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdClicked");
    [_channel invokeMethod:@"onAdClicked" arguments:nil];
}

/**
 *  banner2.0广告点击以后即将弹出全屏广告页
 */
- (void)unifiedBannerViewWillPresentFullScreenModal:(GDTUnifiedBannerView *)unifiedBannerView
{
    
}

/**
 *  banner2.0广告点击以后弹出全屏广告页完毕
 */
- (void)unifiedBannerViewDidPresentFullScreenModal:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdOpenOverlay");
    [_channel invokeMethod:@"onAdOpenOverlay" arguments:nil];
}

/**
 *  全屏广告页即将被关闭
 */
- (void)unifiedBannerViewWillDismissFullScreenModal:(GDTUnifiedBannerView *)unifiedBannerView
{
    
}

/**
 *  全屏广告页已经被关闭
 */
- (void)unifiedBannerViewDidDismissFullScreenModal:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdCloseOverlay");
    [_channel invokeMethod:@"onAdCloseOverlay" arguments:nil];
}

/**
 *  当点击应用下载或者广告调用系统程序打开
 */
- (void)unifiedBannerViewWillLeaveApplication:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdLeftApplication");
    [_channel invokeMethod:@"onAdLeftApplication" arguments:nil];
}

/**
 *  banner2.0被用户关闭时调用
 */
- (void)unifiedBannerViewWillClose:(GDTUnifiedBannerView *)unifiedBannerView
{
    NSLog(@"onAdClosed");
    [_channel invokeMethod:@"onAdClosed" arguments:nil];
    self -> bv = nil;
}
@end

@implementation FlutterUnifiedBannerViewFactory {
    NSObject<FlutterBinaryMessenger>* _messenger;
}

- (instancetype)init:(NSObject<FlutterBinaryMessenger>*) messenger
{
    self = [super init];
    if (self) {
        _messenger = messenger;
    }
    return self;
}

- (nonnull NSObject<FlutterPlatformView> *)createWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args {
    return [[FlutterUnifiedBannerView alloc] initWithFrame:frame viewId:viewId arguments:args binaryMessenger:_messenger];
}

- (NSObject<FlutterMessageCodec>*)createArgsCodec {
    return FlutterStandardMessageCodec.sharedInstance;
}

@end

File: ios/Classes/FlutterUnifiedBannerView.swift
//
//  FlutterUnifiedBannerView.swift
//  adnet_qq
//
//  Created by Dolphin on 2020/12/11.
//

import Foundation
import Flutter

public class FlutterUnifiedBannerView:NSObject, FlutterPlatformView, GDTUnifiedBannerViewDelegate {
    private let channel: FlutterMethodChannel
    private let viewId:Int64
    private let frame: CGRect
    private let posId: String
    private var bv: GDTUnifiedBannerView?

    init(frame: CGRect, viewId: Int64, args: [String: Any], messeneger: FlutterBinaryMessenger) {
        self.viewId = viewId
        self.posId = args["posId"] as! String
        self.frame = frame
        self.channel = FlutterMethodChannel(name: "\(UNIFIED_BANNER_VIEW_ID)_\(viewId)", binaryMessenger: messeneger)
        super.init()
        self.channel.setMethodCallHandler { [weak self] (flutterMethodCall: FlutterMethodCall, flutterResult: FlutterResult) in self?.onMethodCall(call: flutterMethodCall, result: flutterResult)}
    }

    public func view() -> UIView {
        return getView()
    }
    
    private func dispose() {
        bv?.removeFromSuperview()
        bv = nil
        channel.setMethodCallHandler(nil)
    }
    
    private func getView() -> GDTUnifiedBannerView {
        if let bv = bv {
            return bv
        }
        let rect = frame.width == 0 ? CGRect(x: 0, y: 0, width: 1, height: 1) : frame
        let bv = GDTUnifiedBannerView.init(frame: rect, placementId: self.posId, viewController: (UIApplication.shared.keyWindow?.rootViewController)!)
        self.bv = bv
        bv.animated = false
        bv.delegate = self
        return bv
    }
    
    private func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
        switch call.method {
        case "refresh":
            self.getView().loadAdAndShow()
            result(true)
        case "close", "dispose":
            self.dispose()
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    /**
     *  请求广告条数据成功后调用
     *  当接收服务器返回的广告数据成功后调用该函数
     */
    public func unifiedBannerViewDidLoad(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        //print("onAdReceived")
        self.channel.invokeMethod("onAdReceived", arguments: nil);
    }

    /**
     *  请求广告条数据失败后调用
     *  当接收服务器返回的广告数据失败后调用该函数
     */
    public func unifiedBannerViewFailed(toLoad:GDTUnifiedBannerView, error: Error)
    {
        //print("onNoAd \(error)")
        self.channel.invokeMethod("onNoAd", arguments:error.localizedDescription);
    }

    /**
     *  banner2.0曝光回调
     */
    public func unifiedBannerViewWillExpose(_ unifiedBannerView: GDTUnifiedBannerView)
    {
        //print("onAdExposure")
        self.channel.invokeMethod("onAdExposure", arguments:nil);
    }

    /**
     *  banner2.0点击回调
     */
    public func unifiedBannerViewClicked(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        //print("onAdClicked")
        self.channel.invokeMethod("onAdClicked", arguments:nil);
    }

    /**
     *  banner2.0广告点击以后即将弹出全屏广告页
     */
    public func unifiedBannerViewWillPresentFullScreenModal(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        
    }

    /**
     *  banner2.0广告点击以后弹出全屏广告页完毕
     */
    public func unifiedBannerViewDidPresentFullScreenModal(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        //print("onAdOpenOverlay")
        self.channel.invokeMethod("onAdOpenOverlay", arguments:nil);
    }

    /**
     *  全屏广告页即将被关闭
     */
    public func unifiedBannerViewWillDismissFullScreenModal(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        
    }

    /**
     *  全屏广告页已经被关闭
     */
    public func unifiedBannerViewDidDismissFullScreenModal(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        //print("onAdCloseOverlay")
        self.channel.invokeMethod("onAdCloseOverlay", arguments:nil);
    }

    /**
     *  当点击应用下载或者广告调用系统程序打开
     */
    public func unifiedBannerViewWillLeaveApplication(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        //print("onAdLeftApplication")
        self.channel.invokeMethod("onAdLeftApplication", arguments:nil)
    }

    /**
     *  banner2.0被用户关闭时调用
     */
    public func unifiedBannerViewWillClose(_ unifiedBannerView:GDTUnifiedBannerView)
    {
        //print("onAdClosed")
        self.dispose()
        self.channel.invokeMethod("onAdClosed", arguments:nil);
        self.bv = nil;
    }
}

public class FlutterUnifiedBannerViewFactory: NSObject, FlutterPlatformViewFactory {
    private let messenger: FlutterBinaryMessenger
    
    init(messeneger: FlutterBinaryMessenger) {
        self.messenger = messeneger
    }

    public func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
            return FlutterUnifiedBannerView(
                frame: frame,
                viewId: viewId,
                args: args as? [String : Any] ?? [:],
                messeneger: messenger
            )
    }

    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}
